### PR TITLE
add more RDF* support in the API

### PR DESCRIPTION
**Pull request by pchampin**
in rio_api
- rename NamedOrBlankNode to GraphName
- add type Subject, which is NamedOrBlankNode + Triple (if 'star' enabled)
- Subject and Term are marked as non_exhaustive enums;
  this is a breaking change, but allows non-star crates (e.g. rio_xml)
  to co-exist with star crates

in other crates:
- propagate the changes above
- in particular, because Subject and Term are now non-exhaustive enums,
  some From implementations have been replaced with TryFrom;
  this should be zero-cost when all existing variants are actually
  handled.

NB: a consequence of adding the Triple variant is that
owned versions of Terms can no longer be converted into the borrowed versions
(or at least not easily).
It appears that such conversions were not really needed
(for example, they were used to compare owned terms with borrowed terms,
which is now achieved by implementing PartialEq instead).

### CODE REVIEW

**Review by Tpt — COMMENTED** [review]
Thank you for this change.

Just a question: I don't understand what is the problem that requires adding `#[non_exhaustive]` to `Term` and `Subjecŧ` that did not appeared in my previous change.
Indeed, not having `non_exhaustive` requires to add `#[cfg(feature="rio_api/rdf-star")]` everywhere `Term` or `Subjecŧ` is matched but I don't see a limitation outside of it. Where am I wrong?

File: api/src/model.rs [hunk 1]
```diff
@@ -241,6 +265,36 @@ impl<'a> fmt::Display for Triple<'a> {
     }
 }
 
+/// A restriction of [Term] that can be used in the graph name position.
+///
+/// The default string formatter is returning an N-Triples, Turtle and SPARQL compatible representation.
+#[derive(Eq, PartialEq, Debug, Clone, Copy, Hash)]
+pub enum GraphName<'a> {
```
[the remark this answers is not in the document]
Comment (pchampin) [reply api/src/model.rs:272]: Yes, I thought about it and tried it, but that's a non-trivial change. In particular, it breaks the expectation, in several parts of the code, that GraphName can be converted to a Term.

This is not RDF*-specific, so I will not focus on this right now.

**Comment by pchampin** [comment]
About `#[non_exhaustive]`, it is indeed a subtle problem, which took me some time to understand and overcome.

Consider a crate `foo` using `rio_turtle` and `rio_xml`. The author of `foo` wants to add support for Turtle*, so they now require the `star` feature on `rio_turtle`: all of a sudden, `rio_xml` does not compile anymore :-(
This is because `rio_turtle+star` forces the `star` feature on `rio_api`, and `rio_xml` depends on `rio_api` *without* dealing with the additional variants introduced by that feature. `star` is (in that case) what we might call a breaking feature.

My first attempt was to add a `star` feature to `rio_xml`, to handle (with errors) the new variants introduced by that feature in `rio_api`, but that's really not a good solution:
* `star` is not really a feature of `rio_xml`, which still does not supports RDF*;
* it forces users (e.g. the author for crate `foo` in the example above) to add the `star` feature to `rio_xml` whenever they add it to `rio_turtle`, otherwise they will still have the compilation error in `rio_xml`;
* the solution above is counter-intuitive, as illustrated in the first argument, and not trivial to find when faced with the compilation errors;
* in general, getting compilation errors in a crate that one is importing is bad UX, IMO.

Now, replace `rio_xml` in the example above with another crate, by a third party (let's call it `rdfa`), which relies on `rio_api` but *did not* introduce anything to support the `star` feature. Now the author of the `foo` crate has no means to fix the compilation error in that third-party crate, so they *can not use* at the same time `rio_turtle+star` and `rdfa`.

To sum it up: the `star` feature will eventually break other crates that do not use/need it. I find it very much preferable to make that break impact Rio's direct users (the author of `rdfa` in the example above) rather than indirect users (the author of `foo`). And I find it preferable to make it happen all the time rather than in some particular situation that may go unnoticed in tests, and is hard to understand and debug when you face it.

An alternative is of course to have a completely parallel set of types (as for `generalized`) and parsers...

**Comment by Tpt** [comment]
Thank you! I see the problem now. If we have to keep having `non_exhaustive`, what about removing the `star` feature in `rio_api` and having the feature always enabled?
RDF* support does not add much code overhead to `rio_api` and it would make `match` simpler when having the `star` feature on while not adding any any more code is `star` is off (people would just have to keep a `_` match). What do you think about it?

**Comment by pchampin** [comment]
My guess is that `non_exhaustive` is zero_cost. I.e., I assume that, although the compiler forces you to write a `_ => ...` clause in any `match` clause, it strips them from the compiled code when it realize that there are actually no other variants. So compiling without the `star` feature would result in faster code, so I would rather keep it optional.

Note however that I didn't check this assumption (I am not fluent enough in assembly language) but that seems consistent with other things that I have read about compiler-optimizations in Rust.

**Comment by Tpt** [comment]
Yes, indeed, `non_exhaustive` should be zero cost at runtime. But I believe it not "not zero cost" if we consider developer time and having to write down not useful `unreachable!` clutters a bit the code.

**Comment by pchampin** [comment]
> But I believe it not "not zero cost" if we consider developer time and having to write down not useful unreachable! clutters a bit the code.

In the current situation, all depending crates will need to add:
```rust
    #[cfg(feature = "star")
    Triple(_) => panic_or_return_error(),
```
just in case another dependency requires the `star` feature (not to mention the "dummy" feature to be added in `Cargo.toml`, and the need to know when to enable it). And this will break again in the future if we introduce yet another feature-gated extension.

With `non_exhaustive` enums, developers need to add
```rust
   _ => panic_or_return_error(),
```
and nothing new in `Cargo.toml`. This will work out of the box, whether or not other dependencies require the `star` feature. And this will avoid future extensions to break this crate.
believe that adding `non_exhaustive` would be good practice. It will avoid future breaks if you later decide to support yet another extension (e.g. N3 variables :wink:).

It does clutter the code a little bit, but not as much as the "breaking feature" option, IMO.

**Comment by Tpt** [comment]
Thank you for moving forward with it!

About the API, indeed, if we keep the `rdf_star` feature in rio_api, having `#[non_exhaustive]` is the way to go. I was arguing about not having an `rdf_star` feature and so always have a `Triple` option. Support for n3 variables is an other good point. Let's keep these enums `#[non_exhaustive]` for now!

Thank you again!

**Comment by pchampin** [comment]
> I was arguing about not having an `rdf_star` feature 

Ah ok. My initial concern was that RDF* support would have a performance overhead in parsers, which would justify to hide this behind a feature gate.

However, I did some tests:
* `cargo bench ntriples/ntriples # in the master branch`
* `cargo bench ntriples/ntriples # in the rdf_star branch`
* `cargo bench ntriples/ntriples --features star # in the rdf_star branch`

and all results seem to be identical. I would be more confident if you reproduced them on another machine, though. But if that's confirmed, I think we can indeed get rid of the feature gate.

**Comment by Tpt** [comment]
>   I would be more confident if you reproduced them on another machine, though. 

I ran the benchmark on my laptop. I also get similar performance with and without RDF* feature.
This is not very surprising, if I read your code correctly, RDF* support adds to the "hot path" only an extra comparison to `'<'` to term parsing.

However, your branch seems slightly slower than master. But I did not get very consistent results (wall time benchmark and laptops do not play well together) so it might be just noise. It could be interesting to have a look at [iai](https://bheisler.github.io/criterion.rs/book/iai/iai.html).

So, having a boolean flag on the parsers to enable/disable RDF* might be a good option. Or leaving the parsers users raising an error when getting a `Term::Triple` variant. This later option might be also good because parsers users would have to deal with this variant anyway to get exhaustive matches.

**Comment by pchampin** [comment]
Re. removing the feature gate: I also agree that the best option is to let users deal with unsupported term kinds, as the non-exhaustive enum forces them to do anyway. This makes the boolean flag unnecessary, IMO.

Re. performances in `master` vs `rdf-star`: it that's close enough to noise, I guess this is acceptable :)

Re. `iai`: gosh, another tool to get familiar with :-/ But thanks for the link, it looks cool indeed.

**Comment by pchampin** [comment]
@Tpt sorry I missed your message last week (I was taking a few days off). You are always welcome to contribute to this PR, of course. This is your project after all :)

It would indeed be nice to go forward with this, especially now that the RDF-star community group is encouraging test reports. This week is really packed, but hopefully I can move forward on this.

**Comment by Tpt** [comment]
@pchampin Thank you! I moved forward last week-end on building a SPARQL parser crate independent from Oxigraph so it's definitely not lost time ;-). If you think you can move forward on this I might more focus on the SPARQL* support. This way we would have quickly a complete implementation. If not, I would be very happy to get these parsers done.

**Comment by pchampin** [comment]
@Tpt sounds like a plan :)

**Comment by Tpt** [comment]
Update on the Oxigraph side: I have implemented SPARQL-star (both query and update) and added to Oxigraph testsuite code what is needed to build testsuite evaluation reports.

**Comment by pchampin** [comment]
@Tpt I finally found some time to work on this again. I rebased this branch onto the most recent `master`.

I also found a desktop computer to run the benchmarks in a more reproducible way than on my laptop. Result: you were right, this PR slows down the `ntriples/ntriples` by ~4%, and that's without enabling the `rdf-star` feature... (enabling it adds a ~1.5% additional slow down).

This is not terrible, but this is not great. What do you think? Reducing the performance cost can be achieved, but at the cost of more redundant code between the "non-star" and the "star" parsers... :-/

**Comment by Tpt** [comment]
> @Tpt I finally found some time to work on this again. I rebased this branch onto the most recent master.

Great! Thank you so much! For a strange reason Github displays merge conflicts. I don't know why.

> This is not terrible, but this is not great. What do you think? Reducing the performance cost can be achieved, but at the cost of more redundant code between the "non-star" and the "star" parsers... :-/

Imho it's fine. The slow down is not very significant. My guess from profiling is that our way of reading from the `BufRead` stream byte by byte is very inefficient and we could get very significant speedup being smarter on this place. The code base is already complicated, let's not complicate it further for a few % of performance.

**Comment by pchampin** [comment]
FTR the Turtle and TriG parser were much less impact by the refactoring than the N-Triples parser, performancewise (which is not very surprising, as they are already more complex): less than 3% of a slow-down.

Only think left is to add support for `<< .. >>` and `{| .. |}` in these parsers.

**Comment by pchampin** [comment]
Turtle-star and TriG-star are now implemented. They pass all the syntax tests, but some eval tests fail.
I believe this is not due to an error in the parser (the output of the tests seem correct), but to the fact that the isomorphism algorithm used to check the tests is not RDF-star-ready...
@Tpt would you mind have a look at it?

**Comment by Tpt** [comment]
>  @Tpt would you mind have a look at it?

Sure. I hope to be able to do it this week-end. I believe the easiest way to work now is to merge this PR into the rdf-star branch. This way I can work on it easily. I have invited you as a "maintainer" of this git repository. This way you will be able to edit the branch directly too. Are you fine with it?

PS: Do you want to be in the "authors" of `rio_api` and `rio_turtle`?

**Comment by pchampin** [comment]
Just accepted the invitation. This will indeed be more convenient :)

> PS: Do you want to be in the "authors" of `rio_api` and `rio_turtle`?

That would be nice, yes! Thanks for proposing.

**Comment by Tpt** [comment]
@pchampin I just updated the branch with isomorphim. all tests seems to pass now. Congrats!

**Comment by pchampin** [comment]
Congrats to you! :) Now we need to generate an EARL report :-P

**Comment by Tpt** [comment]
> Now we need to generate an EARL report :-P

Yes, it's my next step. I hope to do so this week-end.
